lib/Sema/CSSimplify.cpp

```diff
@@ -1962,21 +1962,32 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
       auto meta1 = cast<AnyMetatypeType>(desugar1);
       auto meta2 = cast<AnyMetatypeType>(desugar2);
 
+      auto instanceType1 = meta1->getInstanceType();
+      auto instanceType2 = meta2->getInstanceType();
+
       ConstraintKind subKind = ConstraintKind::Equal;
       // A.Type < B.Type if A < B and both A and B are classes.
-      if (isa<MetatypeType>(meta1) &&
-          meta1->getInstanceType()->mayHaveSuperclass() &&
-          meta2->getInstanceType()->getClassOrBoundGenericClass())
-        subKind = std::min(kind, ConstraintKind::Subtype);
-      // P.Type < Q.Type if P < Q, both P and Q are protocols, and P.Type
-      // and Q.Type are both existential metatypes.
-      else if (isa<ExistentialMetatypeType>(meta1))
+      if (isa<MetatypeType>(meta1)) {
+        // Let's wait to resolve this relation until
+        // both underlying types are known because in
+        // case when inheritance involved we need to
+        // check subtype relation instead of equality.
+        if (instanceType1->isTypeVariableOrMember() ||
```

Comment:
I think this is reasonable to check only for metatypes in this cause because existentials are going to form `subtype` anyway, WDYT?

Reply:
ok dah, there is an easier way to do this, I'm going to force push in a sec.